platform/engineering/request-for-comment/2024-09-23-evidence-collection-and-polling.md

```diff
@@ -0,0 +1,136 @@
+
+For some documents, this is required, such as 21-4142, since it requires manual action be done by a person. This is also appropriate for documents which fail to go the primary/Documents API path (due to error, data issues/discrepancies, downtimes, network issues, etc)
+
+
+#### Unauthenticated User Document Uploads
+As of now we are not sure the unauthenticated document upload should be any different that authenitcated. We still have a duty to ensure that the document gets where it was going, and potentially a duty to notify the uploader if we cannot.
+
+
+#### Other considerations
+All documents that cannot go their primary path (Documents API) should have a backup path/failsafe by attempting submission via the benefits intake API, unless your primary path is the benefits API or something that uses the benefits API downstream.
+
+Any document that exhausts all sending options, requires action on the systems part to notify the user of this failure, prompting the user to reattempt some other way.
+
+Monitoring of each and every total-failure/exhaustion event should be setup, with investigation and remediation of the error casuing the failure to be a top priority. Remediation of the CAUSE of the error is to prevent the same error from affecting future users, this is not the same as resending or reattempting after failure, since that will not be acceptable after failure emails start going out, as it will cause duplicates to be sent. (TBD on if this is true or not, duplicates may be better than nothing if the user does not act on the email to reattempt, this is a VA business choice).
+
+
+### Post-transmittal validation and status checking
+
+Lighthouse only supports async acceptance of documents, given that, as a steward of the info we accept, we must rely on polling (or an event bus or webhook, if one comes into existence) to keep track of what we send to the downstream systems. 
+
+Failure should be defined by either 
+* Inability to get to the downstream system: refusal (or inability due to an error) to accept the file from the downstream service (Lighthouse in most cases here). In this case, we can either choose to immediately determine that this failure cannot be resolved and move onto the next step in the process, or we can retry some number of times in an attempt to resubmit. 
+* Downstream Issues: Acceptance by 1 or more downstream system, but the file erroring, or not reaching the system of record, in some predefined amount of time.
+
+Exhaustion can be determined by systematic polling and maintaining of statuses via a state system/machine, or database records.  
+We should maintain the state of every document we send via polling, until it reaches a final state. 
+We should poll until the state represents that the document made it into the system of record.
```

Comment:
I think we should poll until the state represents finality. So we poll until either success or unrecoverable error has occurred.

Reply:
For cases where something is stuck in a state, should we keep polling forever? I am inclined to say maybe, as that may force whatever system is holding it up, or where it is stuck to be investigated? Either way we will need to support "after some amount of time, send the user a notification we have not yet been able to deliver your doc".

Reply:
Agreed. I think the statement after this in the doc covers that. There needs to be some sort of "time out" but it may be form dependent. If we wait 14 days then ask the veteran to resubmit, but on day 15 their document is accepted, it could have an impact on what they are trying to do.

Reply:
The can of worms I'm trying to open, because I really dislike doc upload being async AND happening after claim establishment while also not being guaranteed to succeed:

- IF BIP Claims Evidence API provides synchronous responses, AND
- IF BIP Claims Evidence API allows evidence upload to happen before and separately from evidence association with a claim ID ([and I believe they do](https://[claimevidence-api-test.dev.bip.va.gov](https://claimevidence-api-test.dev.bip.va.gov/api/v1/rest/swagger-ui.html#/Association/associate)/api/v1/rest/swagger-ui.html#/Association/associate)], 
- THEN would the evidence upload experience be less faulty and less of a pain to monitor if we redesigned the evidence submission experience so that we make sure all evidence given by a veteran uploads successfully before establishing the claim? Then, for example, if an evidence upload fails, our web UX will tell them what failed and why to give them a chance to fix/reupload their evidence or we provide a path that says "I understand I'm submitting my claim without the following evidence: [list of file names]". We could either do this as they're pressing submit or immediately as they're uploading that piece of evidence (probably this latter option preferred for tighter feedback loop).

I'm leaving separate the question here of whether we'd use BIP directly or whether we'd ask LH to expose these capabilities to us.

Reply:
Do both, benefits intake API and document upload API require pooling?